README.md

* [RFC7230 - HTTP/1.1: Message Syntax and Routing - low-level message parsing and connection management](http://tools.ietf.org/html/rfc7230)
* [RFC7231 - HTTP/1.1: Semantics and Content - methods, status codes and headers](http://tools.ietf.org/html/rfc7231)
* [RFC7232 - HTTP/1.1: Conditional Requests - e.g., If-Modified-Since](http://tools.ietf.org/html/rfc7232)
* [RFC7233 - HTTP/1.1: Range Requests - getting partial content](http://tools.ietf.org/html/rfc7233)
* [RFC7234 - HTTP/1.1: Caching - browser and intermediary caches](http://tools.ietf.org/html/rfc7234)
* [RFC7235 - HTTP/1.1: Authentication - a framework for HTTP authentication](http://tools.ietf.org/html/rfc7235)

* https://github.com/KeepSafe/aiohttp/blob/master/aiohttp/client.py
* https://github.com/python/cpython/blob/master/Lib/http/client.py
* https://github.com/joyent/node/blob/master/lib/_http_client.js
* https://github.com/lukasa/hyper - SPDY, HTTP2
* https://idea.popcount.org/2014-04-03-bind-before-connect/

# Features
* Early error support ([HTTP/1.1 8.2.3](http://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.2.3) `Expect: 100-continue`)
* Mid-stream error support
* Chunked i/o, event support
* Zero-length chunk keepalive
* asyncio.IncompleReadError.expected is the total expected size, not the remaining size


Changes.rst
Version 0.0.1 (2014-08-04)
--------------------------
- Begin development.


t.py
#!/usr/bin/env python3.4

import asyncio, requests
from random import random

class WatClient(asyncio.Protocol):
    def __init__(self, method, resource):
        self.method = method
        self.resource = resource

    def __call__(self):
        return self

    def connection_made(self, transport):
        transport.write("{} {}\n\n".format(self.method, self.resource).encode())
#        print('data sent: {}'.format(self.message))

    def data_received(self, data):
        print('data received: {}'.format(data.decode()))

    def connection_lost(self, exc):
        print('server closed the connection')
#        asyncio.get_event_loop().stop()


@asyncio.coroutine
def h():
    return requests.get("https://google.com")

@asyncio.coroutine
def g(i):
    print("Begin", i)
    result = yield from h()
    print("End", i, result)

@asyncio.coroutine
def f():
    for i in range(10):
        yield from g(i)

loop = asyncio.get_event_loop()
coro = loop.create_connection(WatClient('GET', '/'), 'localhost', 9000)
loop.run_until_complete(coro)
loop.run_forever()
loop.close()


httpnext/__init__.py
from __future__ import print_function, unicode_literals

from asyncio import get_event_loop, Protocol, sleep, coroutine, Future

# Should this inherit from both asyncio.protocol and HTTPConnection? (Probably not, but might be OK)

class HTTPConnection(Protocol):
    def __init__(self, host, port):
        self.host = host
        self.port = port
        self._loop = get_event_loop()
        self._conn = None

    def connect(self):
        self._conn = self._loop.create_connection(lambda: self, self.host, self.port)
        self._loop.run_until_complete(self._conn)

    def close(self):
        pass
        #self._conn.close()

#    @coroutine
#    def driver(self):
#        while not self._eof_received:
#            print("Waiting...")
#            yield from sleep(1)
#        print("Done")

    def request(self, method, url, body=None, headers={}, callback=None):
        if method == "POST":
            headers["Expect"] = "100-continue"
#            headers["Content-Length"] = str(len(body))
            headers["Transfer-Encoding"] = "chunked"
        self._ready_to_send_body, self._eof_received = False, False
        self.body = body
        self.response = b""
        self.transport.write("{} {} HTTP/1.1\r\n".format(method, url).encode())
        for header in headers:
            self.transport.write("{}:{}\r\n".format(header, headers[header]).encode())
        self.transport.write(b"\r\n")
        self._ready_to_send_body = Future()
        self._loop.run_until_complete(self._ready_to_send_body)
        self._loop.run_until_complete(self._send_body())
        self._done = Future()
        if callback is None:
            self._loop.run_until_complete(self._done)
            return self.response
        else:
            pass

    def getresponse(self):
        pass

    def __del__(self):
        self.close()

    def connection_made(self, transport):
        self.transport = transport
        print("Connection made:", transport)

    def connection_lost(self, exc):
        print('server closed the connection')
        #asyncio.get_event_loop().stop()

    def eof_received(self):
        print("Received EOF")
        self._done.set_result(None)
#        self._eof_received = True

    def pause_writing(self):
        print("Received pause writing cb")

    def resume_writing(self):
        print("Received resume writing cb")

    def data_received(self, data):
        print("Received data:", data)
        # TODO: fix this up to not be stupid
        if data.startswith(b"HTTP/1.1 100 Continue"):
            print("Got 100 continue, ready to send body")
            self._ready_to_send_body.set_result(None)
        self.response += data

    @coroutine
    def _send_body(self):
        print("Will send body")
        from io import BytesIO
        b = BytesIO(self.body)
        cs = 1024
        while True:
            c = b.read(cs)
            if c == b"":
                break
            yield from sleep(0.1)
            yield from self._send_chunk(c)

        self.transport.write(b"0\r\n\r\n")
        print("Done writing body")

    @coroutine
    def _send_chunk(self, chunk):
        self.transport.write(bytes("{:x}\r\n".format(len(chunk)), encoding="utf-8"))
        self.transport.write(chunk)
        self.transport.write(b"\r\n")
        print("Wrote", len(chunk), "bytes")

c = HTTPConnection("localhost", 9000)
c.connect()
response = c.request("POST", "/", headers={"foo": "bar", "wat": "wat"}, body=b"0123456789ABCDEF"*1024)
print("Got response:", response)

class HTTPResponse(object):
    def __iter__(self):
        # Chunked response iterator
        pass

    def __enter__(self):
        pass
    def __exit__(self):
        pass

    def read(self, amt=None):
        pass
    def readinto(self, b):
        pass
